### PR TITLE
3.x Upgrade NVidia CUDA toolkit from 11.4.4 to 11.7.1

### PR DESCRIPTION
**Pull request by davprat**
### Description of changes
* Upgraded NVIDA CUDA Toolkit to version 11.7.1 in response to [CVE](https://nvd.nist.gov/vuln/detail/CVE-2022-21821)

### Tests
* Modified [integration test](https://github.com/aws/aws-parallelcluster/pull/4331) to be CUDA version agnostic.
* Modified AMI [validation](https://github.com/aws/aws-parallelcluster/pull/4305) test to support new separate samples package on ARM architecture.

### References
* Requires this [PR](https://github.com/aws/aws-parallelcluster/pull/4382)
* [CUDA 11.7.1 Release Notes](https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html)

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: attributes/default.rb [hunk 1]
```diff
@@ -185,11 +185,13 @@
 # NVIDIA
 default['cluster']['nvidia']['enabled'] = 'no'
 default['cluster']['nvidia']['driver_version'] = '470.141.03'
-default['cluster']['nvidia']['cuda_version'] = '11.4'
+default['cluster']['nvidia']['cuda_version'] = '11.7'
+default['cluster']['nvidia']['cuda_sample_version'] = '11.6'
```
Comment (gmarciani) [inline attributes/default.rb]: Minor: the package is `cuda-samples`. What about using the same name in dictionary key, i.e. `cuda_samples_version` and `cuda_samples_url`?

File: attributes/default.rb [hunk 2]
```diff
@@ -185,11 +185,13 @@
...
 default['cluster']['nvidia']['driver_url_architecture_id'] = arm_instance? ? 'aarch64' : 'x86_64'
 default['cluster']['nvidia']['cuda_url_architecture_id'] = arm_instance? ? 'linux_sbsa' : 'linux'
 default['cluster']['nvidia']['driver_url'] = "https://us.download.nvidia.com/tesla/#{node['cluster']['nvidia']['driver_version']}/NVIDIA-Linux-#{node['cluster']['nvidia']['driver_url_architecture_id']}-#{node['cluster']['nvidia']['driver_version']}.run"
-default['cluster']['nvidia']['cuda_url'] = "https://developer.download.nvidia.com/compute/cuda/11.4.4/local_installers/cuda_11.4.4_470.82.01_#{node['cluster']['nvidia']['cuda_url_architecture_id']}.run"
+default['cluster']['nvidia']['cuda_url'] = "https://developer.download.nvidia.com/compute/cuda/11.7.1/local_installers/cuda_11.7.1_515.65.01_#{node['cluster']['nvidia']['cuda_url_architecture_id']}.run"
+default['cluster']['nvidia']['sample_url'] = "https://github.com/NVIDIA/cuda-samples/archive/refs/tags/v#{node['cluster']['nvidia']['cuda_sample_version']}.tar.gz"
```
Comment (gmarciani) [inline attributes/default.rb]: The main goal of this PR seems to be the upgrade to CUDA 11.7.1. Why should we include NVIDIA Samples?

File: cookbooks/aws-parallelcluster-install/recipes/nvidia.rb [hunk 3]
```diff
@@ -83,6 +83,30 @@
...
   end
 
+  # Get CUDA Sample Files
+  cuda_tmp_samplefile = "/tmp/cuda-sample.tar.gz"
+  remote_file cuda_tmp_samplefile do
+    source node['cluster']['nvidia']['sample_url']
+    mode '0644'
+    retries 3
+    retry_delay 5
+    not_if { ::File.exist?("/usr/local/cuda-#{node['cluster']['nvidia']['cuda_version']}/samples") }
```
Comment (gmarciani) [inline cookbooks/aws-parallelcluster-install/recipes/nvidia.rb]: This path is built twice (here and in line 107). What about having a single declaration and reuse the variable?

File: cookbooks/aws-parallelcluster-install/recipes/nvidia.rb [hunk 4]
```diff
@@ -83,6 +83,30 @@
...
+
+  # Unpack CUDA Samples
+  bash 'cuda.sample install' do
+    user 'root'
+    group 'root'
+    cwd '/tmp'
+    code <<-CUDA
+      set -e
+      tar xf #{cuda_tmp_samplefile}
+      mv "./cuda-samples-#{node['cluster']['nvidia']['cuda_sample_version']}" "/usr/local/"
```
Comment (gmarciani) [inline cookbooks/aws-parallelcluster-install/recipes/nvidia.rb]: Minor: could be done in one line `tar xf #{cuda_tmp_samplefile} --directory "/usr/local/"`?

File: attributes/default.rb [hunk 5]
```diff
@@ -185,11 +185,13 @@
 # NVIDIA
 default['cluster']['nvidia']['enabled'] = 'no'
 default['cluster']['nvidia']['driver_version'] = '470.141.03'
-default['cluster']['nvidia']['cuda_version'] = '11.4'
+default['cluster']['nvidia']['cuda_version'] = '11.7'
```
Comment (gmarciani) [inline attributes/default.rb:192]: Did you check that CUDA 11.7 .1 does not contain any breaking change, according to its changelog?
In any case, may you please link the changelog into the PR description?

File: attributes/default.rb (hunk 2 above)
Comment (davprat) [reply attributes/default.rb]: The samples contain a utility that we use to validate the installation.

File: attributes/default.rb (hunk 5 above)
Comment (davprat) [reply attributes/default.rb:192]: Link added.

File: cookbooks/aws-parallelcluster-install/recipes/nvidia.rb (hunk 3 above)
Comment (davprat) [reply cookbooks/aws-parallelcluster-install/recipes/nvidia.rb]: Fixed

File: cookbooks/aws-parallelcluster-install/recipes/nvidia.rb (hunk 4 above)
Comment (davprat) [reply cookbooks/aws-parallelcluster-install/recipes/nvidia.rb]: Fixed

File: attributes/default.rb (hunk 1 above)
Comment (davprat) [reply attributes/default.rb]: Fixed.

File: attributes/default.rb (hunk 2 above)
Comment (gmarciani) [reply attributes/default.rb]: Got it thanks

**Review by gmarciani — APPROVED** [review]

**Pull request merged**